# Robinhood API Tutorial

1. [Setting up your development environment](#Setting-up-your-development-environment)
2. [Logging in and getting started]
3. [Basic demo program]()
4. [Exploring the API](#Exploring-the-API)
5. Getting data
6. Implementing an automated trading strategy

## Setting up your development environment

We'll be using the [robin-stocks unofficial Robinhood API](https://github.com/jmfernandes/robin_stocksh) to build Python programs that interface with your Robinhood account. You'll need to be 18+ to open an account. This tutorial assumes you know some Python basics and already have Python3 set up. If you need a syntax refresher check out this [cheat sheet](https://www.codecademy.com/learn/learn-python-3/modules/learn-python3-syntax/cheatsheet) from Codecademy. There are plenty of tutorials available for setting up Python on your OS, so that will not be covered here. Keep in mind that standard market hours are 6:30AM - 1:00PM PST Mon-Fri.

First, we need to install the API:

In [1]:
pip install robin_stocks

     |████████████████████████████████| 52 kB 46 kB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


**Note**: your output will look different, as I'm running this in a [Jupyter Notebook](https://jupyter.org/index.html).

## Logging in and getting started

In [2]:
import robin_stocks as rh

user = "example@gmail.com"
pswd = "example"

rh.login(user, pswd)

Enter Robinhood code for validation: 917248


{'access_token': 'eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJleHAiOjE1OTUwMjAzMDMsInRva2VuIjoiVldqNU92ZEl2SDgwNkdEY2EzWXg3Y2lUZHZHQ1BEIiwidXNlcl9pZCI6IjhhZjQ3YzcwLTdmMWYtNDQyZi1hNTEwLTBhODVhMmZkYzU2NiIsImRldmljZV9oYXNoIjoiYjM1NjBjODE0MGQ2ZDllYzIyNGFiNmVjZGZlZDQ5YjAiLCJzY29wZSI6ImludGVybmFsIiwidXNlcl9vcmlnaW4iOiJVUyIsIm9wdGlvbnMiOnRydWUsImxldmVsMl9hY2Nlc3MiOmZhbHNlfQ.c00fqL0QJyqZuCnhrDpcQkg-ZVpxpqf2Rpo1OwYkHGLtkYPpSj8_9TgxErav0NCeGiP-Pt8qNpELEilvCkvF3H_fwtGCMotrkSdgd8OHILdJ06NvDL4LZt1YJjPf2JIfwEoWMhHCySyTgZLT1CPyLphHJfYdHnatNXsDcbQ-txl9HelvkSt_RMuGXVJGUByYF0yKFm_ZjCEAzbk_mQdqxEEih6wn2d3wkY30RfP5SqsjV29tgr1ym9Bg2fAnNeAp2M-4YwTlFvZ4AEvX61QmeYdWSaMyLO2LBO3PX594MLROL3YLYcuAwn9YfnuIhW73t8NwKr-sizNlC_eP885W_w',
 'expires_in': 419490,
 'token_type': 'Bearer',
 'scope': 'internal',
 'refresh_token': 'IDMupkJxrATRVe8dSwMoeX4DOpIrZd',
 'mfa_code': None,
 'backup_code': None,
 'detail': 'logged in with brand new authentication code.'}

Notice we were prompted to enter in a validation code. Robinhood sent me an SMS to authenticate the login before I can use the API. Then it generated an access token so that we can stay logged in without authenticating every time we acess the API in the next 419490 ms (7 min). We will delve into this process further later on. 

For now, we've sucessfully logged in and are ready to start using the API commands. Let's make sure everything is working by getting Apple's current pricing information:

In [21]:
rh.stocks.get_quotes("TSLA")

[{'ask_price': '1547.000000',
  'ask_size': 12,
  'bid_price': '1491.480000',
  'bid_size': 1,
  'last_trade_price': '1544.650000',
  'last_extended_hours_trade_price': '1542.000000',
  'previous_close': '1394.280000',
  'adjusted_previous_close': '1394.280000',
  'previous_close_date': '2020-07-09',
  'symbol': 'TSLA',
  'trading_halted': False,
  'has_traded': True,
  'last_trade_price_source': 'consolidated',
  'updated_at': '2020-07-11T00:00:00Z',
  'instrument': 'https://api.robinhood.com/instruments/e39ed23a-7bd1-4587-b060-71988d9ef483/'}]

It looks like the key `last_trade_price` gives us the current stock price during trading hours. If we want to only access this information, we can specify with the `info` modifier: 

In [20]:
rh.stocks.get_quotes("TSLA", info="last_trade_price")

['1544.650000']

I opened Robinhood the app to double check that the output matches the actual current stock price. They matched, so we are good to go! 

Note: the info modifer works for all the commands that output a list like the one above.

## Basic Demo Program

We are ready to make a simple program that demonstrates some of the API's capabilities. 

Psuedo code:

```
login
output Tesla stock price
check if Tesla stock has fractional shares trading enabled
    if yes:
        buy $1 of $TSLA
        sell when it goes up by 1%
```

First I looked for a way check if a stock is fractionally tradeable on Robinhood. `stocks.get_quotes()` gave us all of the pricing info of a stock, but what about other non-pricing related information? BoOm - `stocks.get_instruments_by_symbols` comes in clutch by dumping out a list of everything else about a stock:

In [31]:
rh.stocks.get_instruments_by_symbols("TSLA")

[{'id': 'e39ed23a-7bd1-4587-b060-71988d9ef483',
  'url': 'https://api.robinhood.com/instruments/e39ed23a-7bd1-4587-b060-71988d9ef483/',
  'quote': 'https://api.robinhood.com/quotes/TSLA/',
  'fundamentals': 'https://api.robinhood.com/fundamentals/TSLA/',
  'splits': 'https://api.robinhood.com/instruments/e39ed23a-7bd1-4587-b060-71988d9ef483/splits/',
  'state': 'active',
  'market': 'https://api.robinhood.com/markets/XNAS/',
  'simple_name': 'Tesla',
  'name': 'Tesla, Inc. Common Stock',
  'tradeable': True,
  'tradability': 'tradable',
  'symbol': 'TSLA',
  'bloomberg_unique': 'EQ0000000003531703',
  'margin_initial_ratio': '0.5200',
  'maintenance_ratio': '0.4000',
  'country': 'US',
  'day_trade_ratio': '0.2500',
  'list_date': '2010-06-29',
  'min_tick_size': None,
  'type': 'stock',
  'tradable_chain_id': '0b29f27a-fbf1-4e11-b887-5597d422957f',
  'rhs_tradability': 'tradable',
  'fractional_tradability': 'tradable',
  'default_collar_fraction': '0.05'}]

The second-to-last item is `fractional_tradability`, which tells us if a stock is fractionally `tradable`. We can access this with the `info` modifier.

In [ ]:
rh.stocks.get_instruments_by_symbols("TSLA", info="fractional_tradability")

Then I found the `orders.order_buy_fractional_by_price` function in the robin_stocks documentation [here](https://robin-stocks.readthedocs.io/en/latest/functions.html#robin_stocks.orders.order_buy_fractional_by_price) so I can buy fractional shares in dollars:

![orders.order_buy_fractional_by_price explanation](assets/fractional$.png)

this takes in the stock symbol (TSLA in our case), how many dollars of the stock we want, and the time aspects of the order.

In [54]:
rh.orders.order_buy_fractional_by_price("TSLA", 1)

{'non_field_errors': ['Invalid time in force for fractional order.']}

My order failed because I executed the command on a Sunday, when the stock market is closed.

Finally, we must keep in mind that all of these commands allow us to input a string (1 stock) or a list of strings (multiple stocks). Because of this, all of them output a list, even when we input 1 stock. Thus, if we want to access the output in a boolean operation, we must access the first element of the returned list.

#### Putting it all together:

In [58]:
# Demo program

import robin_stocks as rh

user = "test@gmail.com"
pswd = "example"

rh.login(user, pswd)

print("Tesla stock price: " + rh.stocks.get_quotes("TSLA", info="last_trade_price")[0]) # accessing the first element of the returned list

if (rh.stocks.get_instruments_by_symbols("TSLA", info="fractional_tradability")[0] == 'tradable'):
    print("Attempting to buy")
    print(rh.orders.order_buy_fractional_by_price("TSLA", 1))
else: 
    print("Tesla isn't fractionally tradeable")

ERROR: There was an issue loading pickle file. Authentication may be expired - logging in normally.
Tesla stock price: 1544.650000
Attempting to buy
{'non_field_errors': ['Invalid time in force for fractional order.']}


## Exploring the API

https://robin-stocks.readthedocs.io/en/latest/functions.html - all functions available in the API are listed in the documentation

# work in progress zone starts now ignore below here

### 2-Factor Authentication

Notice how this will prompt 2-factor authentication.
This means the user has to input a login code sent by Robinhood via SMS/email every time the login function is executed. This would be very annoying if setting up an automated trading system. 
We can get around this by enabling 2 Factor Authentication and logging in with an authentication code. To do this, log in to Robinhood and go to settings, then scroll down and enable 2FA, then choose Authentication App, click "Can't Scan It?" and copy the code. (see GIF)

![settings](assets/2FA.GIF)


In [5]:
from pyrh import Robinhood

user = "test@gmail.com"
pswd = "test"

qr = "test"

rh = Robinhood()
rh.login(username = user, password = pswd, qr_code = qr)
rh.print_quote("AAPL")